src/main.rs

```diff
@@ -1,3 +1,204 @@
-fn main() {
-    println!("Hello, world!");
+use dashmap::DashMap;
+use ropey::Rope;
+use tower_lsp::{
+    jsonrpc::Result,
+    lsp_types::{
+        Diagnostic, DiagnosticSeverity, DidChangeTextDocumentParams, DidOpenTextDocumentParams,
+        InitializeParams, InitializeResult, InitializedParams, NumberOrString, Position, Range,
+        ServerCapabilities, TextDocumentSyncCapability, TextDocumentSyncKind, Url,
+    },
+    Client, LanguageServer, LspService, Server,
+};
+use tree_sitter_lint::{
+    tree_sitter::{self, InputEdit, Parser, Point, Tree},
+    tree_sitter_grep::{Parseable, SupportedLanguage},
+};
+
+#[derive(Debug)]
+struct Backend {
+    client: Client,
+    per_file: DashMap<Url, PerFileState>,
+}
+
+impl Backend {
+    pub fn new(client: Client) -> Self {
+        Self {
+            client,
+            per_file: Default::default(),
+        }
+    }
+
+    async fn run_linting_and_report_diagnostics(&self, uri: &Url) {
+        let per_file_state = self.per_file.get(uri).unwrap();
+        let violations = tree_sitter_lint_local::run_for_slice(
+            &per_file_state.contents,
+            Some(&per_file_state.tree),
+            "dummy_path",
```

Comment:
Will probably want to figure out how to convert a `Url` to a "real" path in case any rules care about it?